File: crates/zou-deno/src/node/zlib.js
// node:zlib, which is gzip, deflate and their inflates in the three
// shapes node offers each of them in: a synchronous call, a call with
// a callback, and a stream.
//
// The compression itself is not here. It is the deflate this runtime
// already links in to open a package tarball, reached through an id a
// job is held by, so a stream can hand it four kilobytes at a time and
// get a single gzip out at the end rather than a hundred of them
// stapled together.
//
// What is missing is missing by name. Brotli is a different algorithm
// and not another header on this one, so `brotliCompress` and the rest
// of that family refuse rather than pretending, and so do the two
// dictionary settings, which zlib takes and this does not pass on.

import { Buffer } from "node:buffer";
import { Transform } from "node:stream";

const ops = Deno.core.ops;

// Node's numbers, which a package passes back to node and never reads.
// They are here because an options object built from them should not
// be full of `undefined`.
export const constants = {
  Z_NO_FLUSH: 0,
  Z_PARTIAL_FLUSH: 1,
  Z_SYNC_FLUSH: 2,
  Z_FULL_FLUSH: 3,
  Z_FINISH: 4,
  Z_BLOCK: 5,
  Z_OK: 0,
  Z_STREAM_END: 1,
  Z_NEED_DICT: 2,
  Z_ERRNO: -1,
  Z_STREAM_ERROR: -2,
  Z_DATA_ERROR: -3,
  Z_MEM_ERROR: -4,
  Z_BUF_ERROR: -5,
  Z_VERSION_ERROR: -6,
  Z_NO_COMPRESSION: 0,
  Z_BEST_SPEED: 1,
  Z_BEST_COMPRESSION: 9,
  Z_DEFAULT_COMPRESSION: -1,
  Z_DEFAULT_STRATEGY: 0,
  Z_FILTERED: 1,
  Z_HUFFMAN_ONLY: 2,
  Z_RLE: 3,
  Z_FIXED: 4,
  DEFLATE: 1,
  INFLATE: 2,
  GZIP: 3,
  GUNZIP: 4,
  DEFLATERAW: 5,
  INFLATERAW: 6,
  UNZIP: 7,
};

/// The bytes a caller handed in, whatever they handed them in as. A
/// string is utf8, which is what node reads one as here.
function bytesOf(input) {
  if (typeof input === "string") return new TextEncoder().encode(input);
  if (input instanceof ArrayBuffer) return new Uint8Array(input);
  if (ArrayBuffer.isView(input)) {
    return new Uint8Array(input.buffer, input.byteOffset, input.byteLength);
  }
  const wrong = new TypeError(
    "The \"buffer\" argument must be of type string or an instance of Buffer, TypedArray, DataView, or ArrayBuffer",
  );
  wrong.code = "ERR_INVALID_ARG_TYPE";
  throw wrong;
}

/// The level an options object asks for, in the one number the op
/// takes. Node's default is the library's default and not nine.
function levelOf(options) {
  const asked = options === null || options === undefined ? undefined : options.level;
  if (asked === undefined || asked === null) return -1;
  const level = Number(asked);
  if (!Number.isInteger(level) || level < -1 || level > 9) {
    const wrong = new RangeError(`The value of "options.level" is out of range. Received ${asked}`);
    wrong.code = "ERR_OUT_OF_RANGE";
    throw wrong;
  }
  return level;
}

/// A failure from the library, in the shape node throws one: a code a
/// caller branches on, and an errno it almost never reads.
function failed(why) {
  const wrong = why instanceof Error ? why : new Error(String(why));
  wrong.code = wrong.code ?? "Z_DATA_ERROR";
  wrong.errno = wrong.errno ?? constants.Z_DATA_ERROR;
  return wrong;
}

function joined(parts) {
  let size = 0;
  for (const part of parts) size += part.length;
  const all = new Uint8Array(size);
  let at = 0;
  for (const part of parts) {
    all.set(part, at);
    at += part.length;
  }
  return Buffer.from(all.buffer, all.byteOffset, all.byteLength);
}

/// One shot: open, write everything, end. This is what every `*Sync`
/// call is, and what the callback ones do before calling back.
function once(kind, input, options) {
  const bytes = bytesOf(input);
  const id = ops.op_zou_zlib_open(kind, levelOf(options));
  try {
    const first = ops.op_zou_zlib_write(id, bytes);
    const last = ops.op_zou_zlib_end(id);
    return joined([first, last]);
  } catch (why) {
    ops.op_zou_zlib_drop(id);
    throw failed(why);
  }
}

/// A transform that is a compression, which is the whole of what every
/// `create*` returns. The job is opened when the first chunk arrives
/// rather than at construction, so a stream nobody writes to and
/// nobody ends does not hold one.
class Zlib extends Transform {
  #kind;
  #level;
  #id = 0;

  constructor(kind, options) {
    super(options);
    this.#kind = kind;
    this.#level = levelOf(options);
    this.bytesWritten = 0;
  }

  #open() {
    if (this.#id === 0) this.#id = ops.op_zou_zlib_open(this.#kind, this.#level);
    return this.#id;
  }

  _transform(chunk, encoding, back) {
    try {
      const bytes = bytesOf(chunk);
      this.bytesWritten += bytes.length;
      const made = ops.op_zou_zlib_write(this.#open(), bytes);
      back(null, made.length > 0 ? Buffer.from(made) : undefined);
    } catch (why) {
      this.#drop();
      back(failed(why));
    }
  }

  _flush(back) {
    try {
      const made = ops.op_zou_zlib_end(this.#open());
      this.#id = 0;
      back(null, made.length > 0 ? Buffer.from(made) : undefined);
    } catch (why) {
      this.#drop();
      back(failed(why));
    }
  }

  #drop() {
    if (this.#id !== 0) {
      ops.op_zou_zlib_drop(this.#id);
      this.#id = 0;
    }
  }

  // Node's own name for it, and the one `pipeline` calls on the way
  // out when something upstream failed.
  destroy(why) {
    this.#drop();
    return super.destroy ? super.destroy(why) : this;
  }

  close(back) {
    this.#drop();
    if (typeof back === "function") queueMicrotask(back);
  }
}

/// The callback form, whose work is synchronous and whose answer is
/// not: node calls back from the thread pool, and a package that reads
/// a variable the callback sets on the next line is broken either way.
function later(kind, input, options, back) {
  if (typeof options === "function") {
    back = options;
    options = {};
  }
  if (typeof back !== "function") {
    const wrong = new TypeError("The \"callback\" argument must be of type function");
    wrong.code = "ERR_INVALID_ARG_TYPE";
    throw wrong;
  }
  let made;
  try {
    made = once(kind, input, options);
  } catch (why) {
    queueMicrotask(() => back(why));
    return;
  }
  queueMicrotask(() => back(null, made));
}

export function gzipSync(input, options) {
  return once("gzip", input, options);
}
export function gunzipSync(input, options) {
  return once("gunzip", input, options);
}
export function deflateSync(input, options) {
  return once("deflate", input, options);
}
export function inflateSync(input, options) {
  return once("inflate", input, options);
}
export function deflateRawSync(input, options) {
  return once("deflateRaw", input, options);
}
export function inflateRawSync(input, options) {
  return once("inflateRaw", input, options);
}
export function unzipSync(input, options) {
  return once("unzip", input, options);
}

export function gzip(input, options, back) {
  later("gzip", input, options, back);
}
export function gunzip(input, options, back) {
  later("gunzip", input, options, back);
}
export function deflate(input, options, back) {
  later("deflate", input, options, back);
}
export function inflate(input, options, back) {
  later("inflate", input, options, back);
}
export function deflateRaw(input, options, back) {
  later("deflateRaw", input, options, back);
}
export function inflateRaw(input, options, back) {
  later("inflateRaw", input, options, back);
}
export function unzip(input, options, back) {
  later("unzip", input, options, back);
}

export function createGzip(options) {
  return new Zlib("gzip", options);
}
export function createGunzip(options) {
  return new Zlib("gunzip", options);
}
export function createDeflate(options) {
  return new Zlib("deflate", options);
}
export function createInflate(options) {
  return new Zlib("inflate", options);
}
export function createDeflateRaw(options) {
  return new Zlib("deflateRaw", options);
}
export function createInflateRaw(options) {
  return new Zlib("inflateRaw", options);
}
export function createUnzip(options) {
  return new Zlib("unzip", options);
}

/// The classes node exports beside the factories. A caller doing
/// `new zlib.Gzip()` gets the same object `createGzip()` returns.
export class Gzip extends Zlib {
  constructor(options) {
    super("gzip", options);
  }
}
export class Gunzip extends Zlib {
  constructor(options) {
    super("gunzip", options);
  }
}
export class Deflate extends Zlib {
  constructor(options) {
    super("deflate", options);
  }
}
export class Inflate extends Zlib {
  constructor(options) {
    super("inflate", options);
  }
}
export class DeflateRaw extends Zlib {
  constructor(options) {
    super("deflateRaw", options);
  }
}
export class InflateRaw extends Zlib {
  constructor(options) {
    super("inflateRaw", options);
  }
}
export class Unzip extends Zlib {
  constructor(options) {
    super("unzip", options);
  }
}

/// Brotli, which this runtime does not have. Every one of these is a
/// refusal by name, because a package that finds `brotliCompress`
/// defined will use it.
function noBrotli(name) {
  return function () {
    throw new Error(`node:zlib ${name} is brotli, which this runtime does not have`);
  };
}

export const brotliCompress = noBrotli("brotliCompress");
export const brotliCompressSync = noBrotli("brotliCompressSync");
export const brotliDecompress = noBrotli("brotliDecompress");
export const brotliDecompressSync = noBrotli("brotliDecompressSync");
export const createBrotliCompress = noBrotli("createBrotliCompress");
export const createBrotliDecompress = noBrotli("createBrotliDecompress");

/// Node's `crc32`, which is the one thing in this module that is not a
/// compression: a checksum over bytes, seeded by whatever the caller
/// had already. It is here because it is cheap and because a package
/// writing its own zip file asks for it.
export function crc32(input, value = 0) {
  const bytes = bytesOf(input);
  let crc = (~value) >>> 0;
  for (let at = 0; at < bytes.length; at++) {
    crc = (crc >>> 8) ^ TABLE[(crc ^ bytes[at]) & 0xff];
  }
  return (~crc) >>> 0;
}

const TABLE = (() => {
  const table = new Uint32Array(256);
  for (let at = 0; at < 256; at++) {
    let value = at;
    for (let bit = 0; bit < 8; bit++) {
      value = value & 1 ? 0xedb88320 ^ (value >>> 1) : value >>> 1;
    }
    table[at] = value >>> 0;
  }
  return table;
})();

// Node hangs the `Z_` numbers off the module as well as off
// `constants`, and a package written before `constants` existed reads
// them there, so the default export carries both.
export default {
  ...constants,
  constants,
  crc32,
  gzip,
  gzipSync,
  gunzip,
  gunzipSync,
  deflate,
  deflateSync,
  inflate,
  inflateSync,
  deflateRaw,
  deflateRawSync,
  inflateRaw,
  inflateRawSync,
  unzip,
  unzipSync,
  createGzip,
  createGunzip,
  createDeflate,
  createInflate,
  createDeflateRaw,
  createInflateRaw,
  createUnzip,
  Gzip,
  Gunzip,
  Deflate,
  Inflate,
  DeflateRaw,
  InflateRaw,
  Unzip,
  brotliCompress,
  brotliCompressSync,
  brotliDecompress,
  brotliDecompressSync,
  createBrotliCompress,
  createBrotliDecompress,
};

File: crates/zou-deno/src/zlib.rs
//! The compression `node:zlib` is a name in front of.
//!
//! A package reaching for zlib is almost always decompressing a body it
//! fetched itself, or gzipping one it is about to send, and both of
//! those are deflate with a different eight or ten bytes around them.
//! The deflate is already linked into this binary, because a package
//! from the registry arrives as a gzipped tarball and something has to
//! open it, so what is here is not a new dependency and not a second
//! implementation. It is that one, with an id javascript can hold.
//!
//! The state is here rather than in the shim because a stream has to
//! keep going: a transform handed four kilobytes at a time cannot
//! compress each of them on its own and call the concatenation a gzip.
//! So a job is opened, written to as many times as the caller likes,
//! and ended, and the one shot calls in the shim are the same three
//! ops with nothing in between them. What comes back from a write is
//! whatever the encoder has finished with by then, which is usually
//! nothing until the window fills, and the rest arrives at the end.
//!
//! Brotli is not here. It is a different algorithm and a different
//! library rather than another header on this one, and `node:zlib`
//! refuses it by name instead of pretending.

use std::collections::HashMap;
use std::io::Write;

use deno_core::{OpState, op2};
use deno_error::JsErrorBox;
use flate2::Compression;
use flate2::write::{
    DeflateDecoder, DeflateEncoder, GzDecoder, GzEncoder, ZlibDecoder, ZlibEncoder,
};

/// How many compression jobs one isolate may hold at once. A stream
/// that is never ended is a job that is never dropped, and this is the
/// same reasoning the socket table's limit has: fail on the box the
/// leak is on rather than take its memory a window at a time.
const OPEN: usize = 256;

/// One compression or decompression in progress, in whichever of the
/// six shapes it was opened as.
///
/// The writer types are the incremental ones: bytes go in with
/// `write_all` and whatever has been finished comes out of the `Vec`
/// underneath, which is drained by every op that looks at it.
enum Job {
    Gzip(GzEncoder<Vec<u8>>),
    Deflate(ZlibEncoder<Vec<u8>>),
    DeflateRaw(DeflateEncoder<Vec<u8>>),
    Gunzip(GzDecoder<Vec<u8>>),
    Inflate(ZlibDecoder<Vec<u8>>),
    InflateRaw(DeflateDecoder<Vec<u8>>),
    /// `unzip`, before enough bytes have arrived to say which of the
    /// two it is. Node's own `unzip` reads the first two bytes and
    /// picks gzip or zlib off them, and so does this.
    Waiting(Vec<u8>),
}

/// Every compression this isolate has open, by the id javascript holds.
#[derive(Default)]
pub struct Jobs {
    last: u32,
    open: HashMap<u32, Job>,
}

impl Job {
    /// The job a name opens, or nothing if this runtime has no such
    /// thing. The names are the shim's and not node's, because node's
    /// are a class name in one place and a function name in another.
    fn named(name: &str, level: Compression) -> Option<Job> {
        Some(match name {
            "gzip" => Job::Gzip(GzEncoder::new(Vec::new(), level)),
            "deflate" => Job::Deflate(ZlibEncoder::new(Vec::new(), level)),
            "deflateRaw" => Job::DeflateRaw(DeflateEncoder::new(Vec::new(), level)),
            "gunzip" => Job::Gunzip(GzDecoder::new(Vec::new())),
            "inflate" => Job::Inflate(ZlibDecoder::new(Vec::new())),
            "inflateRaw" => Job::InflateRaw(DeflateDecoder::new(Vec::new())),
            "unzip" => Job::Waiting(Vec::new()),
            _ => return None,
        })
    }

    fn write(&mut self, bytes: &[u8]) -> std::io::Result<()> {
        match self {
            Job::Gzip(it) => it.write_all(bytes),
            Job::Deflate(it) => it.write_all(bytes),
            Job::DeflateRaw(it) => it.write_all(bytes),
            Job::Gunzip(it) => it.write_all(bytes),
            Job::Inflate(it) => it.write_all(bytes),
            Job::InflateRaw(it) => it.write_all(bytes),
            Job::Waiting(held) => {
                held.extend_from_slice(bytes);
                Ok(())
            }
        }
    }

    /// Whatever is finished, taken out of the buffer underneath so the
    /// same bytes do not come back on the next call.
    fn drained(&mut self) -> Vec<u8> {
        let made = match self {
            Job::Gzip(it) => it.get_mut(),
            Job::Deflate(it) => it.get_mut(),
            Job::DeflateRaw(it) => it.get_mut(),
            Job::Gunzip(it) => it.get_mut(),
            Job::Inflate(it) => it.get_mut(),
            Job::InflateRaw(it) => it.get_mut(),
            Job::Waiting(_) => return Vec::new(),
        };
        std::mem::take(made)
    }

    fn finish(self) -> std::io::Result<Vec<u8>> {
        match self {
            Job::Gzip(it) => it.finish(),
            Job::Deflate(it) => it.finish(),
            Job::DeflateRaw(it) => it.finish(),
            Job::Gunzip(it) => it.finish(),
            Job::Inflate(it) => it.finish(),
            Job::InflateRaw(it) => it.finish(),
            // Nothing ever arrived, or one byte did, which is not
            // enough of a header to be either format. An empty input
            // decompresses to nothing rather than to a complaint,
            // which is what node answers for it too.
            Job::Waiting(held) => match held.is_empty() {
                true => Ok(Vec::new()),
                false => Err(std::io::Error::other("incorrect header check")),
            },
        }
    }
}

/// The two bytes at the front of a gzip member, which is what tells it
/// apart from a zlib stream: no zlib header starts `1f 8b`, because the
/// low nibble of its first byte is the compression method and 8 is the
/// only one there is.
fn sniff(held: &[u8]) -> Option<Job> {
    match held.len() >= 2 {
        false => None,
        true => {
            let mut job = match held[0] == 0x1f && held[1] == 0x8b {
                true => Job::Gunzip(GzDecoder::new(Vec::new())),
                false => Job::Inflate(ZlibDecoder::new(Vec::new())),
            };
            // The held bytes have not been decompressed yet, so they go
            // through the job that was just chosen for them. A failure
            // here is a failure of the next write, which is where the
            // caller is standing.
            let _ = job.write(held);
            Some(job)
        }
    }
}

/// A job, and the id the shim holds it by.
#[op2(fast)]
#[smi]
pub fn op_zou_zlib_open(
    state: &mut OpState,
    #[string] kind: &str,
    #[smi] level: i32,
) -> Result<u32, JsErrorBox> {
    // Node's -1 is "whatever the library thinks", and every other
    // number outside the range is the caller's mistake rather than
    // something to round into place.
    let level = match level {
        -1 => Compression::default(),
        0..=9 => Compression::new(level as u32),
        _ => {
            return Err(JsErrorBox::type_error(format!(
                "{level} is not a compression level"
            )));
        }
    };
    let job = Job::named(kind, level).ok_or_else(|| {
        JsErrorBox::type_error(format!("{kind} is not a compression this runtime has"))
    })?;
    let jobs = state.borrow_mut::<Jobs>();
    if jobs.open.len() >= OPEN {
        return Err(JsErrorBox::type_error(format!(
            "this isolate has {OPEN} compressions open at once, which is all it may have"
        )));
    }
    jobs.last += 1;
    let id = jobs.last;
    jobs.open.insert(id, job);
    Ok(id)
}

/// Bytes in, and whatever came out on the way.
#[op2]
#[buffer]
pub fn op_zou_zlib_write(
    state: &mut OpState,
    #[smi] id: u32,
    #[buffer] bytes: &[u8],
) -> Result<Vec<u8>, JsErrorBox> {
    let jobs = state.borrow_mut::<Jobs>();
    let job = jobs.open.get_mut(&id).ok_or_else(gone)?;
    if let Job::Waiting(held) = job {
        let mut held = std::mem::take(held);
        held.extend_from_slice(bytes);
        match sniff(&held) {
            Some(chosen) => *job = chosen,
            None => {
                *job = Job::Waiting(held);
                return Ok(Vec::new());
            }
        }
    } else {
        job.write(bytes).map_err(|e| failed(&e))?;
    }
    Ok(job.drained())
}

/// The end of the input, which is where a gzip's trailer is written and
/// where a truncated one is noticed. The job is gone either way.
#[op2]
#[buffer]
pub fn op_zou_zlib_end(state: &mut OpState, #[smi] id: u32) -> Result<Vec<u8>, JsErrorBox> {
    let jobs = state.borrow_mut::<Jobs>();
    let job = jobs.open.remove(&id).ok_or_else(gone)?;
    job.finish().map_err(|e| failed(&e))
}

/// A stream that was destroyed rather than ended, which is a job to
/// drop and nothing to say about it.
#[op2(fast)]
pub fn op_zou_zlib_drop(state: &mut OpState, #[smi] id: u32) {
    state.borrow_mut::<Jobs>().open.remove(&id);
}

fn gone() -> JsErrorBox {
    JsErrorBox::type_error("this compression has already been ended")
}

/// What the library says went wrong, which for a corrupt input is a
/// sentence like node's own because both of them come from zlib.
fn failed(why: &std::io::Error) -> JsErrorBox {
    JsErrorBox::type_error(why.to_string())
}
